src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD001UseSwitchToMainThreadAsyncAnalyzer.cs

```diff
@@ -32,26 +27,41 @@ public override void Initialize(AnalysisContext context)
             context.EnableConcurrentExecution();
             context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
 
-            context.RegisterSyntaxNodeAction(Utils.DebuggableWrapper(this.AnalyzeInvocation), SyntaxKind.InvocationExpression);
+            context.RegisterCompilationStartAction(compilationStartContext =>
+            {
+                var legacyThreadSwitchingMembers = CommonInterest.ReadMethods(compilationStartContext.Options, CommonInterest.FileNamePatternForLegacyThreadSwitchingMembers, compilationStartContext.CancellationToken).ToImmutableArray();
+                var analyzer = new Analyzer(legacyThreadSwitchingMembers);
+                compilationStartContext.RegisterSyntaxNodeAction(Utils.DebuggableWrapper(analyzer.AnalyzeInvocation), SyntaxKind.InvocationExpression);
+            });
         }
 
-        private void AnalyzeInvocation(SyntaxNodeAnalysisContext context)
+        private class Analyzer
         {
-            var invocationSyntax = (InvocationExpressionSyntax)context.Node;
-            var invokeMethod = context.SemanticModel.GetSymbolInfo(context.Node).Symbol as IMethodSymbol;
-            if (invokeMethod != null)
+            private readonly ImmutableArray<CommonInterest.QualifiedMember> legacyThreadSwitchingMembers;
+
+            public Analyzer(ImmutableArray<CommonInterest.QualifiedMember> legacyThreadSwitchingMembers)
```

Comment:
nit: I prefer `internal` over `public` for members of non-public types. That way when I'm reviewing code like this and see `public`, I don't have to ask myself if it is *really* public, warranting a deeper consideration for the API.

Reply:
Submitted as #371